mlflow/gateway/client.py

```diff
@@ -110,6 +110,18 @@ def search_routes(self, page_token: Optional[str] = None) -> PagedList[Route]:
         response_json = self._call_endpoint(
             "GET", MLFLOW_GATEWAY_CRUD_ROUTE_BASE, json_body=json.dumps(request_parameters)
         ).json()
+
+        # NB: avoiding a comprehension so as not to double the memory allocation
```

Comment:
Can you elaborate on this?

Reply:
Well, not 'double' per se, but the memory allocation for a shallow copy and the additional object mapping hash table entry is unnecessary overhead for the FastAPI server. I'd like to keep all of these operations as lean as possible.

Reply:
What would the comprehension code look like?

Reply:
```
response_json["routes"] = [
    {**resp, "route_url": assemble_url(self._gateway_uri, resp["name"])} for resp in response_json["routes"]
]
```

I think this is much harder to read as well

Reply:
How much memory allocation can this save?

Reply:
Here's a profiler report. It's not a whole lot (~100kB per call), but the leaner we can make this server's backend processing, the better, in my opinion:

``` python
from concurrent.futures import ProcessPoolExecutor
import copy
from memory_profiler import profile

update_value = "http://127.0.0.1:5000/api/2.0/gateway/routes/conversation-endpoint"

@profile
def modify_dict_for_loop(dict_list):
    for dictionary in dict_list:
        dictionary["key_to_modify"] = update_value
    return dict_list

@profile
def modify_dict_comprehension(dict_list):
    return [{**dictionary, "key_to_modify": update_value} for dictionary in dict_list]

def run_tests():
    # Create a list of 100 dictionaries, each with 10 key-value pairs
    dict_list = [{"key"+str(i): "value"+str(i) for i in range(10)} for _ in range(10000)]
    # Create a copy of the list for the second test
    dict_list_copy = copy.deepcopy(dict_list)

    print("Memory usage for modifying list of dictionaries with for loop:")
    modify_dict_for_loop(dict_list)

    print("\nMemory usage for modifying list of dictionaries with comprehension:")
    modify_dict_comprehension(dict_list_copy)

# Protect the "entry point" of the program
if __name__ == "__main__":
    # Run the tests 10 times in parallel
    with ProcessPoolExecutor(max_workers=10) as executor:
        futures = [executor.submit(run_tests) for _ in range(10)]

    # Wait for all tasks to complete
    for future in futures:
        future.result()
```

``` sh
Filename: /Users/benjamin.wilson/Library/Application Support/JetBrains/PyCharmCE2022.3/scratches/scratch_2.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     7     66.6 MiB     66.6 MiB           1   @profile
     8                                         def modify_dict_for_loop(dict_list):
     9     72.2 MiB      0.0 MiB       10001       for dictionary in dict_list:
    10     72.2 MiB      5.6 MiB       10000           dictionary["key_to_modify"] = update_value
    11     72.2 MiB      0.0 MiB           1       return dict_list



Memory usage for modifying list of dictionaries with comprehension:
Filename: /Users/benjamin.wilson/Library/Application Support/JetBrains/PyCharmCE2022.3/scratches/scratch_2.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     7     66.2 MiB     66.2 MiB           1   @profile
     8                                         def modify_dict_for_loop(dict_list):
     9     71.7 MiB      0.0 MiB       10001       for dictionary in dict_list:
    10     71.7 MiB      5.6 MiB       10000           dictionary["key_to_modify"] = update_value
    11     71.7 MiB      0.0 MiB           1       return dict_list



Memory usage for modifying list of dictionaries with comprehension:
Filename: /Users/benjamin.wilson/Library/Application Support/JetBrains/PyCharmCE2022.3/scratches/scratch_2.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     7     66.4 MiB     66.4 MiB           1   @profile
     8                                         def modify_dict_for_loop(dict_list):
     9     71.9 MiB      0.0 MiB       10001       for dictionary in dict_list:
    10     71.9 MiB      5.5 MiB       10000           dictionary["key_to_modify"] = update_value
    11     71.9 MiB      0.0 MiB           1       return dict_list



Memory usage for modifying list of dictionaries with comprehension:
Filename: /Users/benjamin.wilson/Library/Application Support/JetBrains/PyCharmCE2022.3/scratches/scratch_2.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     7     66.2 MiB     66.2 MiB           1   @profile
     8                                         def modify_dict_for_loop(dict_list):
     9     71.7 MiB      0.0 MiB       10001       for dictionary in dict_list:
    10     71.7 MiB      5.5 MiB       10000           dictionary["key_to_modify"] = update_value
    11     71.7 MiB      0.0 MiB           1       return dict_list



Memory usage for modifying list of dictionaries with comprehension:
Filename: /Users/benjamin.wilson/Library/Application Support/JetBrains/PyCharmCE2022.3/scratches/scratch_2.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     7     66.4 MiB     66.4 MiB           1   @profile
     8                                         def modify_dict_for_loop(dict_list):
     9     71.9 MiB      0.0 MiB       10001       for dictionary in dict_list:
    10     71.9 MiB      5.5 MiB       10000           dictionary["key_to_modify"] = update_value
    11     71.9 MiB      0.0 MiB           1       return dict_list



Memory usage for modifying list of dictionaries with comprehension:
Filename: /Users/benjamin.wilson/Library/Application Support/JetBrains/PyCharmCE2022.3/scratches/scratch_2.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     7     66.5 MiB     66.5 MiB           1   @profile
     8                                         def modify_dict_for_loop(dict_list):
     9     72.0 MiB      0.0 MiB       10001       for dictionary in dict_list:
    10     72.0 MiB      5.5 MiB       10000           dictionary["key_to_modify"] = update_value
    11     72.0 MiB      0.0 MiB           1       return dict_list



Memory usage for modifying list of dictionaries with comprehension:
Filename: /Users/benjamin.wilson/Library/Application Support/JetBrains/PyCharmCE2022.3/scratches/scratch_2.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     7     66.6 MiB     66.6 MiB           1   @profile
     8                                         def modify_dict_for_loop(dict_list):
     9     72.1 MiB      0.0 MiB       10001       for dictionary in dict_list:
    10     72.1 MiB      5.5 MiB       10000           dictionary["key_to_modify"] = update_value
    11     72.1 MiB      0.0 MiB           1       return dict_list



Memory usage for modifying list of dictionaries with comprehension:
Filename: /Users/benjamin.wilson/Library/Application Support/JetBrains/PyCharmCE2022.3/scratches/scratch_2.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     7     65.8 MiB     65.8 MiB           1   @profile
     8                                         def modify_dict_for_loop(dict_list):
     9     71.3 MiB      0.0 MiB       10001       for dictionary in dict_list:
    10     71.3 MiB      5.5 MiB       10000           dictionary["key_to_modify"] = update_value
    11     71.3 MiB      0.0 MiB           1       return dict_list



Memory usage for modifying list of dictionaries with comprehension:
Filename: /Users/benjamin.wilson/Library/Application Support/JetBrains/PyCharmCE2022.3/scratches/scratch_2.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     7     66.0 MiB     66.0 MiB           1   @profile
     8                                         def modify_dict_for_loop(dict_list):
     9     71.5 MiB      0.0 MiB       10001       for dictionary in dict_list:
    10     71.5 MiB      5.5 MiB       10000           dictionary["key_to_modify"] = update_value
    11     71.5 MiB      0.0 MiB           1       return dict_list



Memory usage for modifying list of dictionaries with comprehension:
Filename: /Users/benjamin.wilson/Library/Application Support/JetBrains/PyCharmCE2022.3/scratches/scratch_2.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     7     66.0 MiB     66.0 MiB           1   @profile
     8                                         def modify_dict_for_loop(dict_list):
     9     71.5 MiB      0.0 MiB       10001       for dictionary in dict_list:
    10     71.5 MiB      5.6 MiB       10000           dictionary["key_to_modify"] = update_value
    11     71.5 MiB      0.0 MiB           1       return dict_list



Memory usage for modifying list of dictionaries with comprehension:
Filename: /Users/benjamin.wilson/Library/Application Support/JetBrains/PyCharmCE2022.3/scratches/scratch_2.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    13     72.2 MiB     72.2 MiB           1   @profile
    14                                         def modify_dict_comprehension(dict_list):
    15     77.7 MiB      5.6 MiB       10003       return [{**dictionary, "key_to_modify": update_value} for dictionary in dict_list]


Filename: /Users/benjamin.wilson/Library/Application Support/JetBrains/PyCharmCE2022.3/scratches/scratch_2.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    13     71.7 MiB     71.7 MiB           1   @profile
    14                                         def modify_dict_comprehension(dict_list):
    15     77.3 MiB      5.6 MiB       10003       return [{**dictionary, "key_to_modify": update_value} for dictionary in dict_list]


Filename: /Users/benjamin.wilson/Library/Application Support/JetBrains/PyCharmCE2022.3/scratches/scratch_2.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    13     71.9 MiB     71.9 MiB           1   @profile
    14                                         def modify_dict_comprehension(dict_list):
    15     77.5 MiB      5.6 MiB       10003       return [{**dictionary, "key_to_modify": update_value} for dictionary in dict_list]


Filename: /Users/benjamin.wilson/Library/Application Support/JetBrains/PyCharmCE2022.3/scratches/scratch_2.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    13     71.7 MiB     71.7 MiB           1   @profile
    14                                         def modify_dict_comprehension(dict_list):
    15     77.3 MiB      5.6 MiB       10003       return [{**dictionary, "key_to_modify": update_value} for dictionary in dict_list]


Filename: /Users/benjamin.wilson/Library/Application Support/JetBrains/PyCharmCE2022.3/scratches/scratch_2.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    13     71.9 MiB     71.9 MiB           1   @profile
    14                                         def modify_dict_comprehension(dict_list):
    15     77.5 MiB      5.6 MiB       10003       return [{**dictionary, "key_to_modify": update_value} for dictionary in dict_list]


Filename: /Users/benjamin.wilson/Library/Application Support/JetBrains/PyCharmCE2022.3/scratches/scratch_2.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    13     72.0 MiB     72.0 MiB           1   @profile
    14                                         def modify_dict_comprehension(dict_list):
    15     77.7 MiB      5.7 MiB       10003       return [{**dictionary, "key_to_modify": update_value} for dictionary in dict_list]


Filename: /Users/benjamin.wilson/Library/Application Support/JetBrains/PyCharmCE2022.3/scratches/scratch_2.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    13     71.5 MiB     71.5 MiB           1   @profile
    14                                         def modify_dict_comprehension(dict_list):
    15     77.1 MiB      5.6 MiB       10003       return [{**dictionary, "key_to_modify": update_value} for dictionary in dict_list]


Filename: /Users/benjamin.wilson/Library/Application Support/JetBrains/PyCharmCE2022.3/scratches/scratch_2.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    13     72.1 MiB     72.1 MiB           1   @profile
    14                                         def modify_dict_comprehension(dict_list):
    15     77.8 MiB      5.6 MiB       10003       return [{**dictionary, "key_to_modify": update_value} for dictionary in dict_list]


Filename: /Users/benjamin.wilson/Library/Application Support/JetBrains/PyCharmCE2022.3/scratches/scratch_2.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    13     71.3 MiB     71.3 MiB           1   @profile
    14                                         def modify_dict_comprehension(dict_list):
    15     76.9 MiB      5.6 MiB       10003       return [{**dictionary, "key_to_modify": update_value} for dictionary in dict_list]


Filename: /Users/benjamin.wilson/Library/Application Support/JetBrains/PyCharmCE2022.3/scratches/scratch_2.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    13     71.5 MiB     71.5 MiB           1   @profile
    14                                         def modify_dict_comprehension(dict_list):
    15     77.1 MiB      5.6 MiB       10003       return [{**dictionary, "key_to_modify": update_value} for dictionary in dict_list]
```